### PR TITLE
beginFrameSubscription bugfix and improvement

### PR DESCRIPTION
**Pull request by brenca**
With the current implementation, every time we were notified about a presentation event we asked for a copy of the `BackingStore`, which sadly caused another presentation event, which caused a landslide of presentation events. To fix this, we now check for the `damaged_rect`, which is empty if we ask for a copy of the `BackingStore`.

I've also added another option to the subscriber that allows the user to only get the damaged area in the `frameBuffer`, potentially increasing performance. Specifying this is oprional, so it shouldn't break currently functioning apps using the API.

### CODE REVIEW

File: atom/browser/api/frame_subscriber.cc [hunk 1]
```diff
@@ -14,8 +15,10 @@ namespace api {
 
 FrameSubscriber::FrameSubscriber(v8::Isolate* isolate,
                                  content::RenderWidgetHostView* view,
-                                 const FrameCaptureCallback& callback)
-    : isolate_(isolate), view_(view), callback_(callback), weak_factory_(this) {
+                                 const FrameCaptureCallback& callback,
+                                 const bool& only_damaged)
```
Comment (zcbenz) [inline atom/browser/api/frame_subscriber.cc]: For primitive types, `bool` is preferred than `const bool&`.

File: atom/browser/api/frame_subscriber.cc [hunk 2]
```diff
@@ -27,21 +30,27 @@ bool FrameSubscriber::ShouldCaptureFrame(
   if (!view_ || !host)
     return false;
 
-  const auto size = view_->GetVisibleViewportSize();
+  if (damage_rect.width() == 0 || damage_rect.height() == 0)
```
Comment (zcbenz) [inline atom/browser/api/frame_subscriber.cc]: We can use `damage_rect.IsEmpty()`.

File: docs/api/web-contents.md [hunk 3]
```diff
@@ -917,19 +917,26 @@ For the `mouseWheel` event, the `event` object also have following properties:
 * `hasPreciseScrollingDeltas` Boolean
 * `canScroll` Boolean
 
-### `webContents.beginFrameSubscription(callback)`
+### `webContents.beginFrameSubscription(callback, onlyDamaged)`
```
Comment (zcbenz) [inline docs/api/web-contents.md]: Current convention for APIs is to put callback as last parameter, can you make the signature of this API look like below?

```js
beginFrameSubscription([onlyDamaged, ]callback)
```

**Comment by miniak** [comment]
I would consider using a better name than `onlyDamaged`, affected areas are usually referred to as dirty not damaged.

**Comment by brenca** [comment]
@miniak Agreed, dirty area is a more common name for the affected areas, but I originally decided to go with how chromium named things in their code: https://cs.chromium.org/chromium/src/content/browser/renderer_host/render_widget_host_view_browsertest.cc?rcl=1466839340&l=257

**Comment by zcbenz** [comment]
Thanks!

**Pull request merged**